### ISSUE TITLE
setcchar/getcchar doesn't handle non-BMP and combining characters

### ISSUE THREAD

**Issue by u3shit** [Bill-Gray/PDCursesMod#213]
Currently `setcchar` only looks at the first `wchar_t` of `wch`, which means that
* On braindead windows, non-BMP characters won't work
* Combining characters are silently ignored.

Right now, it looks like only `waddch` is capable of handling combining characters. But `waddch` in case of pdcurses expects Unicode codepoints, while with ncurses it expects whatever weird 8-bit encoding your user set up, so it's not portable either (and one could argue that ncurses is broken if you want to display anything besides ASCII characters).

`getcchar` is similarly broken:
* On windows, non-BMP characters get truncated to 16-bits
* Combining characters will end up as codepoints 0x110001+ due to how pdcurses internally handles combining characters. An application in this case at least can call `PDC_expand_combined_characters` to get back something meaningful but it's not portable.

I'm not sure about windows, generally they use UTF-16, so non-BMP characters should probably end up as a surrogate pairs.

TL;DR I want to display unicode text, help!

**Comment by Bill-Gray** [comment]
(WARNING : it may be a while before I have time to work on any of this!  But by coincidence,  I was looking at just this chunk of code recently and pondering how to tackle the problem.  I should at least write down how it would be done...)

   The comments in `pdcurses/util.c` say (correctly) that combining characters aren't handled,  and that therefore,  "the length returned by `getcchar()` is always 1 or 0."  True for PDCurses,  but it really shouldn't be for PDCursesMod.

   "...On braindead windows, non-BMP characters won't work..." As you say,  that ought to be (basically only can be) handled with surrogate pairs on Microsoft's products.  The `wch` passed to `setcchar()` would be a two-character string,  a “high surrogate” in the 0xD800–0xDBFF range followed by a “low surrogate” in the 0xDC00–0xDFFF range.  (Plus zero or more combining characters,  which might themselves be surrogate pairs.)

So our first step would be to convert the `wchar_t *wch` string into,  say,  a `wint_t *wint` string in which surrogate pairs would be splunged together into SMP chars.  (And we'll need an inverse function to that for `getcchar()`,  but we'll get to that later.)

The resulting `wint_t *wint` string may be of length 1.  In that case,  we store it and we're done.  Otherwise,  there are combining characters involved (*) and we add them to our list,  if needed,  and (as you note) it'll end up as a codepoint 0x110001 or greater,  corresponding to PDCursesMods' internal handling (which is PDCursesMod-specific and which nobody really ought to know about if they can help it.)

On the flip side,  `getcchar()` ought to look at the value stored;  if it's >= 0x110001,  it can call `PDC_expand_combined_characters()` and get back a `wint_t` array.  Otherwise,  it just stores the value into a `wint_t` array of length 1.

If we're on Windows (more specifically,  if `wchar_t` is only 16 bits long),  we go through said array for characters in the SMP (codepoints beyond 0xffff);  if we find them,  we replace them with surrogate pairs.  Either the resulting array of `wchar_t` values is returned,  or its length.

I've just put together [this partial implementation of the above (adds SMP,  but no combining chars yet)](https://www.projectpluto.com/temp/util.c).  Drop it in place of your existing `pdcurses/util.c`,  and I think you'll be partway there.

WARNING : Testing is nearly zero (essentially,  'it compiled on my Linux box with gcc and cross-compiled with MinGW').  At present,  `getcchar()` breaks SMP chars into surrogate pairs even if `sizeof( wchar_t) >= 4`. None of the demos exercises `setcchar()/getcchar()`,  and those functions are not used elsewhere in PDCursesMod.  You'll be an alpha tester here.

**Comment by u3shit** [comment]
Thanks for looking into it! 
Right now my primary development platform is Linux+ncurses and it will take some time until it turns into something releaseable, so it's not urgent, but I'll likely need this on windows where is no normal terminal emulator before 10. (And I'll need to add extra routines to convert my UTF-32 text to UTF-16, just so PDCursesMod can convert it back to basically UTF-32...)

Couldn't you just change `if( c >= 0x10000)` to `if( sizeof( wchar_t) == 2 && c >= 0x10000)` or the likes? (I have no idea what to do on platforms where wchar_t is not 16 or 32 bits, or when CHAR_BIT != 8, is there even a platform like that supported by PDCurses/PDCursesMod?) Gcc will optimize that constant away. Of course, this is only part of the story, combining characters are the meat of the matter here.

**Comment by Bill-Gray** [comment]
"...And I'll need to add extra routines to convert my UTF-32 text to UTF-16, just so PDCursesMod can convert it back to basically UTF-32.."

   And PDCursesMod will need such routines as well.  I've just [modified the 'modified' `util.c`](https://www.projectpluto.com/temp/util.c) to include the necessary `_wchar_to_int32_array()` and `_int32_to_wchar_array()` functions.  Currently static,  but they could be exported.  (I considered using `wint_t` instead of `int32_t`,  but bizarrely,  `wint_t` is 16 bits on Microsoft platforms.)

   At present,  I have `SHORT_WCHAR_T` #defined if we're in Windows;  other similarly benighted platforms may have to be added.  On such platforms,  `_int32_to_wchar_array()` has to expand SMP characters into surrogate pairs.  Sadly,  there's no way to conditionally compile with something as perfectly sensible-looking as

`#if sizeof( wchar_t) == 2`

   I suspect the only `wchar_t` cases we'll encounter are 16,  32,  and 64-bits.  Only the 16-bit ones require this oddball handling.  If we ever encounter 16-bit `char`s,  I think most of PDCurses and PDCursesMod will break.

**Comment by Bill-Gray** [comment]
I think we're good on this now,  and I'll close the issue (feel free to reopen it if you see continuing problems,  of course.)

The recent commits address getting `getcchar()` and `setcchar()` to work with combining and SMP characters.  I added a new `widetest` demo to exercise them.  It appears to be working Just Fine in VT,  WinGUI,  and the Linux framebuffer ports (the only ports currently supporting combining characters and SMP.)

I also made some improvements in handling surrogate pairs.  For example,  that treble clef character in `widetest` was done with a string literal of `L"\xd83d\xdd1e"`,  and works on Linux and Windows(R).  A string literal of `L"\x1d11e"` would work on non-Windows systems only.

`widetest` works in ncurses,  except the treble clef only shows up if you use `L"\x1d11e"`.  I think I'll file an inquiry on the ncurses-bugs list as to whether that's the intended behavior.

**Comment by u3shit** [comment]
Thanks, I'll look into it (as soon as I get my program back into a working condition).
